### PR TITLE
Gaussian cluster distance

### PR DESCRIPTION
**Pull request by Njan0**
# Description

Added evidenceMatchingPDF as distance-function between Gaussian clusters.
Visualize the distance-function over time by outputting a graph as csv-files.

### CODE REVIEW

File: examples/md-flexible/scripts/extractClusterGraphCombine.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
+#!/usr/bin/python3
+
```
Comment (FG-TUM) [inline examples/md-flexible/scripts/extractClusterGraphCombine.py:2]: please document your scripts in the README in this folder: How to use them and the minimal python version that is needed to run (no need to determine the exact, just as low as you know)

File: src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h [hunk 2]
```diff
@@ -206,6 +217,11 @@ bool BayesianClusterSearch::tune(bool currentInvalid) {
     // select best config
     _currentConfig = _encoder.convertFromCluster(_gaussianCluster.getEvidenceMax());
     AutoPasLog(debug, "Selected Configuration {}", _currentConfig.toString());
+
+    // print graph of distances
+    size_t sampleSize = _cellSizeFactors->isFinite() ? _cellSizeFactors->size() : _predNumLHSamples;
+    auto continuousSamples = FeatureVector::lhsSampleFeatureContinuous(sampleSize, _rng, *_cellSizeFactors);
+    _gaussianCluster.logDebugGraph(_neighbourFun, continuousSamples);
```
Comment (FG-TUM) [inline src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h:224]: this function should probably only be called if `loglevel <= debug` .

File: src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h [hunk 3]
```diff
@@ -180,49 +204,84 @@ class GaussianCluster {
...
+   * @return all discrete-continuous tuples paired with their corresponding acquisition
+   */
+  [[nodiscard]] std::vector<VectorAcquisition> sampleAcquisition(
+      AcquisitionFunctionOption af, const std::function<std::vector<VectorDiscrete>(VectorDiscrete)> &neighbourFun,
+      const std::vector<VectorContinuous> &continuousSamples) const {
+    std::vector<VectorAcquisition> acquisitions;
+    // pair up all discrete with all continuous samples
+    acquisitions.reserve(_clusters.size() * continuousSamples.size());
+
+    auto graphStream = logGraphInit();
```
Comment (FG-TUM) [inline src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h:221]: The graph stuff here should only be done if `loglevel <= debug` .

File: src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h [hunk 4]
```diff
@@ -342,70 +406,181 @@ class GaussianCluster {
...
 
-      // get acquisition considering neighbours
-      VectorDiscrete currentDiscrete = Eigen::VectorXi::Zero(_dimRestriction.size());
-      for (size_t i = 0; i < _clusters.size(); ++i, discreteIncrement(currentDiscrete)) {
-        auto neighbours = neighbourFun(currentDiscrete);
+  /**
+   * Create and initialize streams for nodes and edges.
+   * @return
+   */
+  GraphStream logGraphInit() const {
```
Comment (FG-TUM) [inline src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h:508]:
```suggestion
  [[nodiscard]] GraphStream logGraphInit() const {
```

File: src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h [hunk 5]
```diff
@@ -32,6 +32,15 @@ class GaussianCluster {
   // store pairs of vectors and corresponding acquisition
   using VectorAcquisition = std::pair<std::pair<VectorDiscrete, VectorContinuous>, double>;
 
+  // function that generate all neighbouring vectors of given vector
+  using NeighbourFunction = std::function<std::vector<VectorDiscrete>(VectorDiscrete)>;
+
+  // for each vector store a vector of each neighborus and corresponding distance
```
Comment (FG-TUM) [inline src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h:38]:
```suggestion
  // for each vector store a vector of all neighbors and the corresponding distance
```
?

File: examples/md-flexible/input/testTuning/MDFlex_Case2.yaml [hunk 6]
```diff
@@ -0,0 +1,24 @@
+selector-strategy                :  Fastest-Absolute-Value
+tuning-strategy                  :  full-Search
+tuning-interval                  :  1000
+tuning-samples                   :  3
+tuning-max-evidence              :  50
+functor                          :  Lennard-Jones (12-6)
+cutoff                           :  2
+deltaT                           :  0
+cell-size                        :  [1.0,1.5,2.0]
+# 792 tuning iterations + 50 to observe choses configuration
```
Comment (FG-TUM) [inline examples/md-flexible/input/testTuning/MDFlex_Case2.yaml:10]:
```suggestion
# 792 tuning iterations + 50 to observe chosen configuration
```

File: src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h (hunk 3 above)
Comment (FG-TUM) [reply src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h:241]: or better trace

File: src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h [hunk 7]
```diff
@@ -342,70 +406,181 @@ class GaussianCluster {
...
+    AutoPasLog(debug, edgeStream.str());
+    AutoPasLog(debug, "GaussianCluster Graph: End");
+  }
+
+  /**
+   * Default function used to convert vectors to readable strings.
+   * @param vec
+   * @return string with format (a,b,...,n) beginning with discrete values.
+   */
+  static std::string defaultVecToString(const VectorPairDiscreteContinuous &vec) {
```
[the remark this answers is not in the document]
Comment (Njan0) [reply src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h]: VectorPairDiscreteContinuous is a pair of Eigen::VectorXi and Eigen::VectorXd. Additionally begin() and end() is currently not implemented in the latest stable version of Eigen.

[the remark this answers is not in the document]
Comment (FG-TUM) [reply src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h]: As just discussed add a `@note` line in the doc of the function stating pretty much what you just said.

File: src/autopas/selectors/FeatureVectorEncoder.h [hunk 8]
```diff
@@ -185,10 +212,126 @@ class FeatureVectorEncoder {
...
+      // initial value
+      auto init = target[i];
+
+      // for each possible value of that dimension
+      for (int x = 0; x < _dimRestrictions[i]; ++x) {
+        // skip initial value
+        if (x != init) {
+          auto neighbour = target;
+          neighbour[i] = x;
+          result.push_back(std::make_pair(std::move(neighbour), 1.));
```
Comment (FG-TUM) [inline src/autopas/selectors/FeatureVectorEncoder.h:287]:
```suggestion
          result.emplace_back(std::move(neighbour), 1.);
```

File: src/autopas/selectors/FeatureVectorEncoder.h [hunk 9]
```diff
@@ -185,10 +212,126 @@ class FeatureVectorEncoder {
...
+          // check if container changed
+          if (i == containerTraversalEstimatorClusterIndex) {
+            auto xContainer = std::get<0>(_containerTraversalEstimatorOptions[x]);
+            if (targetContainer != xContainer) {
+              // assign lower weight
+              weight = 0.5;
+            }
+          }
+
+          result.push_back(std::make_pair(std::move(neighbour), weight));
```
Comment (FG-TUM) [inline src/autopas/selectors/FeatureVectorEncoder.h:330]:
```suggestion
          result.emplace_back(std::move(neighbour), weight);
```

File: src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h [hunk 10]
```diff
@@ -180,96 +201,133 @@ class GaussianCluster {
...
+    return acquisitions;
+  }
+
+  /**
+   * Generate all possible combinations of discrete tuples and continuous tuples in samples
+   * and calculate their weight to each other. Output graph in AutoPasLog Trace.
+   * @param neighbourFun function which generates neighbours of given discrete tuple
+   * @param continuousSamples continuous tuples
+   */
+  void logDebugGraph(const GaussianModelTypes::NeighbourFunction neighbourFun,
```
Comment (FG-TUM) [inline src/autopas/selectors/tuningStrategy/GaussianModel/GaussianCluster.h:259]:
```suggestion
  void logDebugGraph(const GaussianModelTypes::NeighbourFunction &neighbourFun,
```

File: src/autopas/selectors/tuningStrategy/GaussianModel/GaussianProcess.h [hunk 11]
```diff
@@ -81,6 +81,14 @@ class GaussianProcess {
    */
   [[nodiscard]] size_t numEvidence() const { return _inputs.size(); }
 
+  /**
+   * Get all currently stored evidence
+   * @return pair of inputs and outputs
+   */
+  [[nodiscard]] const std::pair<std::vector<Vector>, Eigen::VectorXd> getEvidence() const {
```
Comment (FG-TUM) [inline src/autopas/selectors/tuningStrategy/GaussianModel/GaussianProcess.h:88]:
```suggestion
  [[nodiscard]] std::pair<std::vector<Vector>, Eigen::VectorXd> getEvidence() const {
```
This `const` does not do anything.

File: src/autopas/selectors/tuningStrategy/GaussianModel/GaussianProcess.h [hunk 12]
```diff
@@ -333,8 +376,16 @@ class GaussianProcess {
...
     }
-    for (auto &hyper : _hypers) {
-      hyper.score /= scoreSum;
+    if (scoreSum > 0) {
+      for (auto &hyper : _hypers) {
+        hyper.score /= scoreSum;
+      }
+    } else {
+      // all scores are 0
+      double uniformProbability = 1 / _hypers.size();
```
Comment (FG-TUM) [inline src/autopas/selectors/tuningStrategy/GaussianModel/GaussianProcess.h:385]:
```suggestion
      double uniformProbability = 1. / _hypers.size();
```

**Pull request merged**